Cache/IOUtils.lean

```diff
@@ -0,0 +1,114 @@
+import Lean.Elab.ParseImportsFast
+import Std.Data.HashMap
+import Std.Data.RBMap
+
+def List.pop : (l : List α) → l ≠ [] → α × List α
+  | a :: as, _ => (a, as)
+
+/-- Removes a parent path from the beginning of a path -/
+def System.FilePath.withoutParent (path : FilePath) (parent : FilePath) : FilePath :=
+  let rec aux : List String → List String → List String
+    | z@(x :: xs), y :: ys => if x == y then aux xs ys else z
+    | [], _ => []
+    | x, [] => x
+  ⟨"/".intercalate $ aux (path.toString.splitOn "/") (parent.toString.splitOn "/")⟩
+
+namespace Cache.IO
+
+open System
+
+/-- Target directory for build files -/
+def LIBDIR : FilePath :=
+  "build" / "lib"
+
+/-- Target directory for caching -/
+def CACHEDIR : FilePath :=
+  ⟨"cache"⟩
+
+/--
+Runs a terminal command and retrieves its output. `↔` must be used as a placeholder for whitespaces
+-/
+def runCmd (cmd : String) : IO String := do
+  let cmd := cmd.splitOn " " |>.map (·.replace "↔" " ")
+  if h : cmd ≠ [] then
+    let (cmd, args) := cmd.pop h
+    let out ← IO.Process.output {
+      cmd  := cmd
+      args := ⟨args⟩
+    }
+    if out.exitCode != 0
+      then throw $ IO.userError out.stderr
+      else return out.stdout
+  else return ""
+
+/-- Spawns a terminal command. `↔` must be used as a placeholder for whitespaces -/
+def spawnCmd (cmd : String) : IO UInt32 := do
+  let cmd := cmd.splitOn " " |>.map (·.replace "↔" " ")
+  if h : cmd ≠ [] then
+    let (cmd, args) := cmd.pop h
+    let child ← IO.Process.spawn {
+      cmd  := cmd
+      args := ⟨args⟩
+    }
+    child.wait
+  else return 0
+
+/-- Recursively gets all files from a directory with a certain extension -/
+partial def getFilesWithExtension
+  (fp : FilePath) (extension : String) (acc : Array FilePath := #[]) :
+    IO $ Array FilePath := do
+  if ← fp.isDir then
+    let mut acc := acc
+    for dirEntry in ← fp.readDir do
+      for innerFp in ← getFilesWithExtension dirEntry.path extension do
+        acc := acc.push innerFp
+    return acc
+  else if fp.extension == some extension then return acc.push fp else return acc
+
+/-- Given a path to a Lean file, concatenates the paths to its build files -/
+def mkBuildPaths (leanPath : FilePath) : String :=
+  " ".intercalate [
+    LIBDIR / leanPath.withExtension "olean" |>.toString,
+    LIBDIR / leanPath.withExtension "ilean" |>.toString,
```

Comment:
```suggestion
    LIBDIR / leanPath.withExtension "ilean" |>.toString,
    LIBDIR / leanPath.withExtension "c" |>.toString,
```

Reply:
Hmm, this adds quite some more complexity to it. Can we ignore `ir` files for now? Or would it trigger recompilation regardless of caching the other files?